egs/wsj/s5/steps/nnet3/multilingual/allocate_multilingual_examples.py

```diff
@@ -198,22 +175,30 @@ def process_multilingual_egs(args):
         eg_to_weight_file_handle = open("{0}/{1}weight.{2}.ark".format(args.egs_dir,
                                                                        args.egs_prefix, archive_index + 1), 'w')
 
-        for round_index in range(num_rounds):
-            # Read 'block_size' examples from each lang and write them to the current output scp file:
-            for lang_index in range(num_langs):
-                example_lines  = read_lines(in_scp_file_handles[lang_index], lang_to_block_size[lang_index])
 
-                for eg_line in example_lines:
-                    eg_id = eg_line.split()[0]
+        for block_index in range(num_blocks_per_this_archive):
+            # find the lang with the highest proportion of remaining examples
+            remaining_proportions = [remain / tot for remain, tot in zip(lang_to_num_remaining_egs, lang_to_num_examples)]
+            lang_index, max_proportion = max(enumerate(remaining_proportions), key=lambda a: a[1])
+
```

Comment:
I think it is still possible to have some archives with no block from small tasks.
Assume that we have two tasks with 100000 and 100 blocks per task and if we used 1 block from task 2 in 1st archive, then it needs to use 1000 blocks from task one to start using another block from 2nd task and if archive size is 100 blocks, then the next 9 archives just contain task 1.

My idea is that in large mismatch condition, it is better to have some boolean option like --allow-unbalanced-epochs, and if it is true, it will copy tasks with smaller data to reach some minimum proportion between smallest to largest task (e.g. 0.1) and print number of extra copies from each task in log.

Reply:
I know, it is possible when there is too much imbalance. I guess it's best if Dan replies to your comment.